Warning: Can't perform a React state update on an unmounted component

I am using `GraphiQL` in the form as a textarea field and every time I hit submit and navigate away from the page I get this warning

`
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
`
   `in GraphiQL (created by GraphQlTextArea)`

Can someone please look into this?

@zoontek Getting these errors now

```
ERROR in ./node_modules/codemirror-graphql/lint.js
Module not found: Error: Can't resolve 'graphql-language-service-interface' in '/.../.../Desktop/.../.../node_modules/codemirror-graphql'
 @ ./node_modules/codemirror-graphql/lint.js 5:39-84
 @ ./node_modules/graphiql/esm/components/QueryEditor.js
 @ ./node_modules/graphiql/esm/components/GraphiQL.js
 @ ./node_modules/graphiql/esm/index.js
...
 @ multi ./main.js webpack-hot-middleware/client

ERROR in ./node_modules/codemirror-graphql/hint.js
Module not found: Error: Can't resolve 'graphql-language-service-interface' in '/.../.../Desktop/.../.../node_modules/codemirror-graphql'
 @ ./node_modules/codemirror-graphql/hint.js 5:39-84
 @ ./node_modules/graphiql/esm/components/QueryEditor.js
 @ ./node_modules/graphiql/esm/components/GraphiQL.js
 @ ./node_modules/graphiql/esm/index.js
...
 @ multi ./main.js webpack-hot-middleware/client
```

--- a/packages/graphiql/src/components/GraphiQL.tsx
+++ b/packages/graphiql/src/components/GraphiQL.tsx
@@ -1001,7 +1001,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
 
   handleEditVariables = (value: string) => {
     this.setState({ variables: value });
-    this._storage.set('variables', value);
+    debounce(500, () => this._storage.set('variables', value));
     if (this.props.onEditVariables) {
       this.props.onEditVariables(value);
     }
@@ -1049,9 +1049,11 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
               this.docExplorerComponent.showDoc(type);
             }
           });
-          this._storage.set(
-            'docExplorerOpen',
-            JSON.stringify(this.state.docExplorerOpen),
+          debounce(500, () =>
+            this._storage.set(
+              'docExplorerOpen',
+              JSON.stringify(this.state.docExplorerOpen),
+            ),
           );
         }
       }
@@ -1114,7 +1116,9 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       const leftSize = moveEvent.clientX - getLeft(editorBar) - offset;
       const rightSize = editorBar.clientWidth - leftSize;
       this.setState({ editorFlex: leftSize / rightSize });
-      this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex));
+      debounce(500, () =>
+        this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex)),
+      );
     };
 
     let onMouseUp: OnMouseUpFn = () => {
@@ -1178,9 +1182,11 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
           docExplorerOpen: true,
           docExplorerWidth: Math.min(docsSize, 650),
         });
-        this._storage.set(
-          'docExplorerWidth',
-          JSON.stringify(this.state.docExplorerWidth),
+        debounce(500, () =>
+          this._storage.set(
+            'docExplorerWidth',
+            JSON.stringify(this.state.docExplorerWidth),
+          ),
         );
       }
       this._storage.set(
@@ -1192,9 +1198,11 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
     let onMouseUp: OnMouseUpFn = () => {
       if (!this.state.docExplorerOpen) {
         this.setState({ docExplorerWidth: hadWidth });
-        this._storage.set(
-          'docExplorerWidth',
-          JSON.stringify(this.state.docExplorerWidth),
+        debounce(500, () =>
+          this._storage.set(
+            'docExplorerWidth',
+            JSON.stringify(this.state.docExplorerWidth),
+          ),
         );
       }
 
@@ -1212,9 +1220,11 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
     this.setState({
       docExplorerWidth: DEFAULT_DOC_EXPLORER_WIDTH,
     });
-    this._storage.set(
-      'docExplorerWidth',
-      JSON.stringify(this.state.docExplorerWidth),
+    debounce(500, () =>
+      this._storage.set(
+        'docExplorerWidth',
+        JSON.stringify(this.state.docExplorerWidth),
+      ),
     );
   };
 
@@ -1249,9 +1259,11 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
           variableEditorHeight: bottomSize,
         });
       }
-      this._storage.set(
-        'variableEditorHeight',
-        JSON.stringify(this.state.variableEditorHeight),
+      debounce(500, () =>
+        this._storage.set(
+          'variableEditorHeight',
+          JSON.stringify(this.state.variableEditorHeight),
+        ),
       );
     };
 

--- a/packages/graphiql/src/components/GraphiQL.tsx
+++ b/packages/graphiql/src/components/GraphiQL.tsx
@@ -241,13 +241,6 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       subscription: null,
       ...queryFacts,
     };
-
-    // Subscribe to the browser window closing, treating it as an unmount.
-    if (typeof window === 'object') {
-      window.addEventListener('beforeunload', () =>
-        this.componentWillUnmount(),
-      );
-    }
   }
 
   componentDidMount() {
@@ -315,7 +308,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
     ) {
       nextSchema = undefined;
     }
-
+    this._storage.set('operationName', nextOperationName as string);
     this.setState(
       {
         schema: nextSchema,
@@ -344,37 +337,6 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       this.variableEditorComponent,
       this.resultComponent,
     ]);
-  }
-
-  // When the component is about to unmount, store any persistable state, such
-  // that when the component is remounted, it will use the last used values.
-  componentWillUnmount() {
-    if (this.state.query) {
-      this._storage.set('query', this.state.query);
-    }
-    if (this.state.variables) {
-      this._storage.set('variables', this.state.variables);
-    }
-    if (this.state.operationName) {
-      this._storage.set('operationName', this.state.operationName);
-    }
-    this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex));
-    this._storage.set(
-      'variableEditorHeight',
-      JSON.stringify(this.state.variableEditorHeight),
-    );
-    this._storage.set(
-      'docExplorerWidth',
-      JSON.stringify(this.state.docExplorerWidth),
-    );
-    this._storage.set(
-      'docExplorerOpen',
-      JSON.stringify(this.state.docExplorerOpen),
-    );
-    this._storage.set(
-      'historyPaneOpen',
-      JSON.stringify(this.state.historyPaneOpen),
-    );
   }
 
   render() {
@@ -832,6 +794,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
         this.docExplorerComponent.showDocForReference(reference);
       }
     });
+    this._storage.set(
+      'docExplorerOpen',
+      JSON.stringify(this.state.docExplorerOpen),
+    );
   };
 
   handleRunQuery = (selectedOperationName?: string) => {
@@ -858,6 +824,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
         response: undefined,
         operationName,
       });
+      this._storage.set('operationName', operationName as string);
 
       if (this._queryHistory) {
         this._queryHistory.updateHistory(editedQuery, variables, operationName);
@@ -979,6 +946,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       query: value,
       ...queryFacts,
     });
+    this._storage.set('query', value);
     if (this.props.onEditQuery) {
       return this.props.onEditQuery(value);
     }
@@ -1033,6 +1001,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
 
   handleEditVariables = (value: string) => {
     this.setState({ variables: value });
+    this._storage.set('variables', value);
     if (this.props.onEditVariables) {
       this.props.onEditVariables(value);
     }
@@ -1080,6 +1049,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
               this.docExplorerComponent.showDoc(type);
             }
           });
+          this._storage.set(
+            'docExplorerOpen',
+            JSON.stringify(this.state.docExplorerOpen),
+          );
         }
       }
     }
@@ -1090,6 +1063,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       this.props.onToggleDocs(!this.state.docExplorerOpen);
     }
     this.setState({ docExplorerOpen: !this.state.docExplorerOpen });
+    this._storage.set(
+      'docExplorerOpen',
+      JSON.stringify(this.state.docExplorerOpen),
+    );
   };
 
   handleToggleHistory = () => {
@@ -1097,6 +1074,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       this.props.onToggleHistory(!this.state.historyPaneOpen);
     }
     this.setState({ historyPaneOpen: !this.state.historyPaneOpen });
+    this._storage.set(
+      'historyPaneOpen',
+      JSON.stringify(this.state.historyPaneOpen),
+    );
   };
 
   handleSelectHistoryQuery = (
@@ -1133,6 +1114,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       const leftSize = moveEvent.clientX - getLeft(editorBar) - offset;
       const rightSize = editorBar.clientWidth - leftSize;
       this.setState({ editorFlex: leftSize / rightSize });
+      this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex));
     };
 
     let onMouseUp: OnMouseUpFn = () => {
@@ -1148,6 +1130,7 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
 
   handleResetResize = () => {
     this.setState({ editorFlex: 1 });
+    this._storage.set('editorFlex', JSON.stringify(this.state.editorFlex));
   };
 
   private _didClickDragBar(event: React.MouseEvent) {
@@ -1195,12 +1178,24 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
           docExplorerOpen: true,
           docExplorerWidth: Math.min(docsSize, 650),
         });
+        this._storage.set(
+          'docExplorerWidth',
+          JSON.stringify(this.state.docExplorerWidth),
+        );
       }
+      this._storage.set(
+        'docExplorerOpen',
+        JSON.stringify(this.state.docExplorerOpen),
+      );
     };
 
     let onMouseUp: OnMouseUpFn = () => {
       if (!this.state.docExplorerOpen) {
         this.setState({ docExplorerWidth: hadWidth });
+        this._storage.set(
+          'docExplorerWidth',
+          JSON.stringify(this.state.docExplorerWidth),
+        );
       }
 
       document.removeEventListener('mousemove', onMouseMove!);
@@ -1217,6 +1212,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
     this.setState({
       docExplorerWidth: DEFAULT_DOC_EXPLORER_WIDTH,
     });
+    this._storage.set(
+      'docExplorerWidth',
+      JSON.stringify(this.state.docExplorerWidth),
+    );
   };
 
   private handleVariableResizeStart: MouseEventHandler<
@@ -1250,6 +1249,10 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
           variableEditorHeight: bottomSize,
         });
       }
+      this._storage.set(
+        'variableEditorHeight',
+        JSON.stringify(this.state.variableEditorHeight),
+      );
     };
 
     let onMouseUp: OnMouseUpFn = () => {

--- a/packages/graphiql/src/components/GraphiQL.tsx
+++ b/packages/graphiql/src/components/GraphiQL.tsx
@@ -1456,8 +1456,8 @@ export class GraphiQL extends React.Component<GraphiQLProps, GraphiQLState> {
       }
       debounce(500, () =>
         this._storage.set(
-          'variableEditorHeight',
-          JSON.stringify(this.state.variableEditorHeight),
+          'secondaryEditorHeight',
+          JSON.stringify(this.state.secondaryEditorHeight),
         ),
       );
     };

